# Exercise 6: numpy

## Aim: 

You can find the teaching resources for this lesson here: https://numpy.org/doc/stable/user/quickstart.html

### Issues covered:
-

## 1. Basics

Let's start by creating some arrays - try to create an array using `a = np.array(1, 2, 3, 4)`. Does this work? Can you edit it to make it work?

In [11]:
a = np.array(1, 2, 3, 4)
# You need to use the square brackets:
a = np.array([1, 2, 3, 4])

TypeError: array() takes from 1 to 2 positional arguments but 4 were given

1. Take a look at the following numpy array:
```
[[7.0, 8.0, 4.0, 2.0],
 [12.0, 1.0, 0.0, 10.0],
 [0.0, 0.0, 0.0, 0.0]]
```
- How many axes does it have?
- What is the length of the array?

Hint: you can use `.ndim` and `.shape` to help if you enclose the array in `np.array()` to define the array.

In [9]:
import numpy as np
array = np.array([[7.0, 8.0, 4.0, 2.0], [12.0, 1.0, 0.0, 10.0], [0.0, 0.0, 0.0, 0.0]])
array.ndim
# There are 2 dimensions! It's a 2D array because it has rows and columns. Even though there are three sets of [ ] brackets, the structure is still 2D because each row is a 1D array and multiple rows together form the 2D array. 
array.shape
# The first dimension (rows) is the outermost brackets - in our case it is 3. The second dimension (columns) is the number of values in each of the inner brackets, which is 4 in our case.

(3, 4)

2. Can you come up with an example of what a 3D array would look like? Use `np.zeros` to make it then print it out and have a look at `.ndim` and `.shape`

In [19]:
import numpy as np
array_3d = np.zeros((3,4,2))
array_3d
array_3d.ndim
array_3d.shape
# A 3D array in numpy can be thought of as a collection of 2D arrays stacked together. Each layer is a 2D array and the third dimension is how many 2D arrays are stacked - in this case we have three lots of 2D arrays with 4 rows and 2 columns

(3, 4, 2)

How many elements are in your array? Use `.size` to check.

In [20]:
array_3d.size

24

What type are the elements in the array? Use `.dtype` to check.

In [21]:
array_3d.dtype

dtype('float64')

How many bites are in each element of the array? Use `.itemsize` to check.

In [22]:
array_3d.itemsize

8

Create a 1D array of numbers 1-10 using `a = np.linspace(1, 10, 10)`. Reshape this to be 2x5 and print it.

In [44]:
a = np.linspace(1, 9, 9).reshape(3,3)
print(a)

[[1. 2. 3.]
 [4. 5. 6.]
 [7. 8. 9.]]


What happens if you multiply your array by 2 using `b = a*2`

In [45]:
b = a*2
# The array has been multiplied by two elementwise (not matrix multiplication)
print(b)

[[ 2.  4.  6.]
 [ 8. 10. 12.]
 [14. 16. 18.]]


How do you do matrix multiplication? Try doing the matrix product of a and b.

In [49]:
# Either of these methods work
matrix_product = a @ b
matrix_product = a.dot(b)
print(matrix_product)

[[ 60.  72.  84.]
 [132. 162. 192.]
 [204. 252. 300.]]


When performing operations between arrays of different data types, numpy automatically converts the result to the more precise type - this is called upcasting. Let's demonstrate this concept:
- Create an array with 3 elements all set to one using `a = np.ones(3, dtype=np.int32)` and set the data type to np.int32
- Create a float array of 3 elements evenly spaced between 0 and pi using `b = np.linspace(0, np.pi, 3)`. The data type will be float64 by default
- Check the data type of both arrays
- Add the arrays a and b to make a new array `c`. Print the resulting array c and its data type.

In [58]:
a = np.ones(3, dtype=np.int32)
b = np.linspace(0, np.pi, 3)
c = a + b 
a.dtype #int32
b.dtype #float64
c.dtype #float64
c

array([1.        , 2.57079633, 4.14159265])

For matrix a in the previous question, what do you think `a.sum()` would be? Check your answer.

In [61]:
# a.sum is the sum of all elements - a unary operation. `.sum()` is a method of the array class because you use it on the array.
a.sum()

3

Create an array using `np.ones(6).reshape(3,2)`. If we only want to sum each column, how would we do that?

In [67]:
a = np.ones(6).reshape(3,2)
a.sum(axis=0)

array([3., 3.])

Create a 1D array of size 20 where each element is the cube of its index. We'll use this array in the next questions.

In [1]:
import numpy as np
a = np.arange(20) ** 3

Print the 5th element of a. Hint: your answer should be 64 - remember where we start indexing.

In [4]:
a[4]

64

Slice the array to get elements from index 3 to index 7 (inclusive).

In [6]:
a[3:8]

array([ 27,  64, 125, 216, 343])

Change every 3rd element to -1. This should give: `[ -1, 1, 8, -1, 64, 125, -1, ... ]`

In [7]:
a[:15:3] = -1
a

array([  -1,    1,    8,   -1,   64,  125,   -1,  343,  512,   -1, 1000,
       1331,   -1, 2197, 2744, 3375, 4096, 4913, 5832, 6859])

Reverse the array and print the result.

In [8]:
reversed_a = a[::-1]
reversed_a

array([6859, 5832, 4913, 4096, 3375, 2744, 2197,   -1, 1331, 1000,   -1,
        512,  343,   -1,  125,   64,   -1,    8,    1,   -1])

Create a 3x4 numpy array `b` using `b = np.array([[2 * i + j for j in range(4)] for i in range(3)])`.

In [11]:
b = np.array([[2 * i + j for j in range(4)] for i in range(3)])
print(b)

[[0 1 2 3]
 [2 3 4 5]
 [4 5 6 7]]


Print the element in the second row and third column. This should be 4.

In [12]:
b[1,2]

4

Extract and print the second column as a 1D array.

In [14]:
second_col = b[:, 1]
second_col

array([1, 3, 5])

Extract and print a sub-array containing the last two rows.

In [15]:
sub_array = b[1:3, :]
sub_array

array([[2, 3, 4, 5],
       [4, 5, 6, 7]])

Use slicing to replace the last row with the values `[7, 7, 7, 7]`.

In [16]:
b[2, :] = [7, 7, 7, 7]
b

array([[0, 1, 2, 3],
       [2, 3, 4, 5],
       [7, 7, 7, 7]])

Iterate over the elements of the array using the `.flat` attribute and print them.

In [17]:
for element in b.flat:
    print(element)

0
1
2
3
2
3
4
5
7
7
7
7


Create a 3D array `c` using `c = np.array([[[i * 10 + j * 5 + k for k in range(4)] for j in range(3)] for i in range(2)])`

In [19]:
c = np.array([[[i * 10 + j * 5 + k for k in range(4)] for j in range(3)] for i in range(2)])
c

array([[[ 0,  1,  2,  3],
        [ 5,  6,  7,  8],
        [10, 11, 12, 13]],

       [[10, 11, 12, 13],
        [15, 16, 17, 18],
        [20, 21, 22, 23]]])

Print all elements of the first layer of the array.

In [21]:
c[0]

array([[ 0,  1,  2,  3],
       [ 5,  6,  7,  8],
       [10, 11, 12, 13]])

Use `...` to print the last element of each 1D array contained within c.

In [22]:
c[..., -1]

array([[ 3,  8, 13],
       [13, 18, 23]])

Modify the first column of the second layer to `[0, 0, 0]`

In [24]:
c[1, :, 0] = [0, 0, 0]
c

array([[[ 0,  1,  2,  3],
        [ 5,  6,  7,  8],
        [10, 11, 12, 13]],

       [[ 0, 11, 12, 13],
        [ 0, 16, 17, 18],
        [ 0, 21, 22, 23]]])

## 2. Shape manipulation

Use the following to create a 3x4 array: 
```
rg = np.random.default_rng(1)
a = np.floor(10 * rg.random((3, 4)))
```

In [10]:
import numpy as np
rg = np.random.default_rng(1)
a = np.floor(10 * rg.random((3, 4)))
a

array([[5., 9., 1., 9.],
       [3., 4., 8., 4.],
       [5., 0., 7., 5.]])

Use the `ravel` method to flatten the array and print it.

In [6]:
a.ravel()

array([5., 9., 1., 9., 3., 4., 8., 4., 5., 0., 7., 5.])

Reshape the array so it has the shape (2,6) and print it.

In [16]:
b = a.reshape(2, 6)
b

array([[5., 9., 1., 9., 3., 4.],
       [8., 4., 5., 0., 7., 5.]])

Transpose the array and print it.

In [17]:
c = a.T
c

array([[5., 8.],
       [9., 4.],
       [1., 5.],
       [9., 0.],
       [3., 7.],
       [4., 5.]])

Use the resize method to change the shape of the array to (6,2). Notice the difference between reshape and resize.

In [18]:
# reshape returns the argument with a modified shape but resize modifies the array itself
a.resize(6,2)
a

array([[5., 9.],
       [1., 9.],
       [3., 4.],
       [8., 4.],
       [5., 0.],
       [7., 5.]])

Reshape the array to a shape of (3, -1) and print the reshaped array. Note what -1 does here.

In [19]:
d =a.reshape(3, -1)
d

array([[5., 9., 1., 9.],
       [3., 4., 8., 4.],
       [5., 0., 7., 5.]])

## 3. Copies and views

## 4. Advanced